system/src/system_connection_manager.cpp

```diff
@@ -0,0 +1,489 @@
+            uint16_t serverPort = 0;
+            switch (a->ai_family) {
+                case AF_INET: {
+                    inet_inet_ntop(a->ai_family, &((sockaddr_in*)a->ai_addr)->sin_addr, serverHost, sizeof(serverHost));
+                    serverPort = ntohs(((sockaddr_in*)a->ai_addr)->sin_port);
+                    break;
+                }
+                case AF_INET6: {
+                    inet_inet_ntop(a->ai_family, &((sockaddr_in6*)a->ai_addr)->sin6_addr, serverHost, sizeof(serverHost));
+                    serverPort = ntohs(((sockaddr_in6*)a->ai_addr)->sin6_port);
+                    break;
+                }
+            }
+            //LOG(TRACE, "test socket=%d, connecting to %s#%u", s, serverHost, serverPort);
+
+            struct ifreq ifr = {};
+            if_index_to_name(connectionMetrics.interface, ifr.ifr_name);
+            r = sock_setsockopt(s, SOL_SOCKET, SO_BINDTODEVICE, &ifr, sizeof(ifr));
+            if (r) {
+                LOG(ERROR, "test socket=%d, failed to sock_setsockopt to IF %s, errno=%d", s, ifr.ifr_name, errno);
+                sock_close(s);
+                return SYSTEM_ERROR_NETWORK;
+            }
+
+            r = sock_connect(s, a->ai_addr, a->ai_addrlen);
+            if (r) {
+                LOG(ERROR, "test socket=%d, failed to connect to %s#%u, errno=%d", s, serverHost, serverPort, errno);
+                return SYSTEM_ERROR_NETWORK;
+            }
+            LOG(INFO, "test socket # %d, %s bound to %s connected to %s#%u", s, ifr.ifr_name, netifToName(connectionMetrics.interface), serverHost, serverPort); 
+
+            // Step 3: Use the socket descriptor for the polling structure, allocate our test buffers
+            connectionMetrics.socketDescriptor = s;
+            pfds[socketCount].fd = connectionMetrics.socketDescriptor;
+            pfds[socketCount].events = (POLLIN | POLLOUT);
+            socketCount++;
+
+            CHECK(allocateTestPacketBuffers(&connectionMetrics));
+        }
+    }
+    
+    // Step 4: Send/Receive data on the sockets for the duration of the test time
+    unsigned packetDataLength = random(1, REACHABILITY_MAX_PAYLOAD_SIZE);
+    auto endTime = HAL_Timer_Get_Milli_Seconds() + REACHABILITY_TEST_DURATION_MS;
+    while (HAL_Timer_Get_Milli_Seconds() < endTime) {
+        CHECK(pollSockets(pfds, socketCount, packetDataLength));
+    }
+
+    // Only read from sockets to receive any final outstanding packets
+    for (int i = 0; i < socketCount; i++) {
+        pfds[i].events = (POLLIN);
+    }
+
+    endTime = HAL_Timer_Get_Milli_Seconds() + REACHABILITY_TEST_DURATION_MS;
+    while(testPacketsOutstanding() && HAL_Timer_Get_Milli_Seconds() < endTime) {
+        pollSockets(pfds, socketCount, packetDataLength);    
+    }
+
+    // Step 5: Cleanup test metrics, close sockets, calculate updated diagnostics
+    testSuccessful = true;
+
+    return 0;
+}
+
+const Vector<ConnectionMetrics>* ConnectionTester::getConnectionMetrics(){
```

Comment:
This is for debug purposes, but again let's avoid returning references/pointers to internal members, especially ones that will cause thread safety issues.